\documentclass{article}
\usepackage[left=1in, right=1in]{geometry}
\usepackage[T1]{fontenc}
\usepackage[cm-default]{fontspec}
\usepackage{lmodern}

\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{sectsty}
\usepackage{ctable}
\usepackage{hyperref}
\usepackage[numbers]{natbib}
\usepackage{subcaption}

\begin{document}

% TODO:
%   * Explain that every SEQC sample was put in the same experiment for
%     isolator.
%
%   * Explain the reasoning behind adding 0.1 tpm to expression values to
%     account for zeros.


\section{Bias Correction}\label{bias-correction}

Isolator attempts to model and correct for multiple factors that can
conflate the transcript expression estimates.

\subsection{3' bias}\label{bias}

Selection of polyadenylated transcripts is a common step mRNA-Seq used
to avoid sequencing introns, partially degraded transcripts, and
ribosomal RNA.

To model the effect we fit a one parameter model in which a transcript
is truncated at any position with probability $p$. The probability of
observing a fragment at a position $k$ of length $\ell$, counting from
the 3' end, is then proportional to the probability that the transcript
was \emph{not} truncated before $k + \ell$, which is $(1-p)^{k + \ell}$.
When fit, $p$ is typically quite small, on the order of $10^{-5}$, in
which case this correction has little effect on transcripts shorter that
several kilobases.

\subsection{Fragmentation effects}\label{fragmentation-effects}

Popular RNA-Seq protocols such as Illumina's TruSeq typically involve a
random fragmentation step. Subtle implications arise from this process.

Existing statisticaly models (see \cite{Pachter:2011wm} for a review),
assume fragments are sampled uniformly at random from a transcript. This
does not exactly match the implications of random fragmentation in
which, under an ideal model, \emph{breakpoints} rather than fragments
are introduced uniformly at random. For a fragment to be observed it
must pass size selection and have fallen between two breakpoints, or one
breakpoint and the end of the transcript. Since the ends of a transcript
act as fixed breakpoints, the result isAsome enrichment of fragments at
either end of the transcript.

\subsection{Fragment GC content}\label{fragment-gc-content}

\section{Analysis}

All reads were aligned using the STAR aligner \cite{Dobin:2013fg}. RNA-Seq
methods vary in their expected input: some work on raw reads, others require
either alignments relative to genome or relative to transcript sequences. STAR
allows the same alignments to be output in transcript and genome coordinates, so
we were able to use the same alignments for each method, excluding the alignment
free methods. Quantification was performed against version 76 of the Ensembl
annotations and revision 38 of the human genome sequence.

\begin{table}
\begin{center}
\begin{tabular}[c]{ll}
\toprule\addlinespace
Method & Version \\
\midrule
STAR & 2.4.0k \\\addlinespace
Cufflinks & 2.2.1 \\\addlinespace
BitSeq & 0.7.0 \\\addlinespace
RSEM & 1.2.19 \\\addlinespace
Sailfish & 0.6.3 \\\addlinespace
eXpress & 1.5.1 \\\addlinespace
Kallisto & 0.42 \\\addlinespace
Salmon & 0.4.1 \\\addlinespace
\bottomrule
\addlinespace
\end{tabular}
\caption{Versions of software used in the analysis.}
\end{center}
\end{table}

\begin{table}
\begin{center}
\begin{tabular}[c]{@{}lrr@{}}
\toprule\addlinespace
Method & $c$ vs $0.75a + 0.25b$ & $d$ vs $0.25a + 0.75b$
\\\addlinespace
\midrule
Isolator & \textbf{0.995} & \textbf{0.995}
\\\addlinespace
BitSeq & 0.994 & 0.994
\\\addlinespace
RSEM/PM & 0.994 & 0.993
\\\addlinespace
RSEM/ML & 0.987 & 0.987
\\\addlinespace
Salmon & 0.986 & 0.986
\\\addlinespace
eXpress & 0.983 & 0.983
\\\addlinespace
Sailfish & 0.978 & 0.976
\\\addlinespace
Kallisto & 0.972 & 0.970
\\\addlinespace
Cufflinks & 0.941 & 0.982
\\\addlinespace
\bottomrule
\addlinespace
\end{tabular}
\caption{Proportionality correlation between gene-level estimates for
the mixed samples C and D and weighted averages of estimates for A and
B, corresponding to the mixture proportions for C and D.}
\end{center}
\label{table:seqcgenes}
\end{table}

\begin{figure}
\begin{center}
\includegraphics[width=0.7\textwidth]{analysis/seqc/runtime.pdf}
\caption{Run time needed to process the SEQC data presented in the Results
sections. All methods were run a Google Compute Engine instance backed by four
Intel Xeon cores and 52GB of a memory.}
\end{center}
\label{fig:runtime}
\end{figure}

\section{Analysis of cardiomyocyte data}

Different methods often produce very different estimates of isoform abundance.
To demonstrate the effect this has on downstream analysis, we analyzed the
cardiomyocyte data from \cite{Kuppusamy:2015ey} using each method compared in
the main paper.

\subsection{qPCR Validation}\label{cardioqpcr}

Total RNAs for 2 independent HFA and HFV samples were extracted using RNAeasy
mini kit (Qiagen) (6). Two independent HAH total RNA samples were purchased
commercially from Clontech and Agilent. Total RNAs for 3 independent replicates
of Day 20 cardiomyocytes (derived from human embryonic stem cells) were
extracted with miRNeasy mini kit (Qiagen). All RNA samples were further DNAse
treated with DNA-free kit (Ambion) and reverse transcription reaction was
carried out using Omniscript RT kit (Qiagen). Gene specific primers were
designed targeting the exon-exon junction for specific exons of interest using
the freely available software APE. The primer sequences were as follows:

\begin{center}
\begin{tabular}{rl}
TNNT2   & F-CAGAGACCATGTCTGACATAGA \\
        & R-CTGCCTCCTCCTGCTCGTx \\
OBSL1   & F-CCTGTACGCTGGTACAAGGA \\
        & R-CTCTTCCACGCTGACAAGGA \\
LTBP4   & F-CAGCCCAAAAAGTGTGCAGG \\
        & R-CTGTCTCTTCCTCACGCTGG \\
OSBPL1A & F-TCACAGCTCTGCTCAACAGG \\
        & R-ACATTGTTTATGGGCCCGGT \\
ACOT9   & F-CCACATTCATGAAGCATGTTCA \\
        & R-CCTACTATCTCCCGCAACT \\
IMMT    & F-TGCACTTTCAGAAGAAGCATCC \\
        & R-TTTTCCTGTTGTGCAAGGCG \\
PALM    & F-TGAAGGCAGCCATGTACTCG \\
        & R-GGACCACTTTGGTCTCGTCC
\end{tabular}
\end{center}

\begin{figure}[H]
\centering
\includegraphics[width=0.2\textwidth]{analysis/cardio/colorkey.pdf} \\
\includegraphics[width=0.9\textwidth]{analysis/cardio/ACOT9-comparison.pdf}
\includegraphics[width=0.9\textwidth]{analysis/cardio/IMMT-comparison.pdf}
\end{figure}
\begin{figure}[H]
\ContinuedFloat
\centering
\includegraphics[width=0.9\textwidth]{analysis/cardio/LTBP4-comparison.pdf}
\includegraphics[width=0.9\textwidth]{analysis/cardio/OBSL1-comparison.pdf}
\includegraphics[width=0.9\textwidth]{analysis/cardio/OSBPL1A-comparison.pdf}
\end{figure}
\begin{figure}[H]
\ContinuedFloat
\centering
\includegraphics[width=0.9\textwidth]{analysis/cardio/PALM-comparison.pdf}
\includegraphics[width=0.9\textwidth]{analysis/cardio/TNNT2-comparison.pdf}
\caption{Point estimates of fold change of exon expression from qPCR and and
RNA-Seq using various methods. Exon expression for RNA-Seq is computed by
estimating transcript expression, then summing the expression of all transcripts
that share a parcitular exon.}
\label{fig:cardiofoldchange}
\end{figure}

\subsection{Principal Component Analysis}\label{cardiopca}

A simple analysis often peformed on gene expression data is principal component
analysis (PCA). Due to large disagreements in isoform abundance estimates, each
method produces qualitatively different results. There is no ground truth for
this data, but the results from Isolator in some ways fit our understanding of
the data more closely than the other methods. Uniquely in the Isolator data,
fetal atrium and fetal ventricle are grouped very closely and clearly separated
from the H7 day20 data.

In most methods the two adult samples are quite heterogenous. The exceptions are
Kallisto and eXpress. While those both show closer agreement between the adult
samples, the fetal samples are far more spread out.

Though Isolator uses a model with informative priors, it is important to note
that the fact that fetal atrium and ventricle tissue in particular are similar
is not something that is encoded is this model.

\begin{figure}[H]
\includegraphics[width=0.2\textwidth]{analysis/cardio/colorkey.pdf} \\
\includegraphics[width=0.5\textwidth]{analysis/cardio/isolator-pca.pdf}
\includegraphics[width=0.5\textwidth]{analysis/cardio/salmon-pca.pdf}
\includegraphics[width=0.5\textwidth]{analysis/cardio/sailfish-pca.pdf}
\includegraphics[width=0.5\textwidth]{analysis/cardio/kallisto-pca.pdf}
\end{figure}
\begin{figure}
\ContinuedFloat
\includegraphics[width=0.5\textwidth]{analysis/cardio/express-pca.pdf}
\includegraphics[width=0.5\textwidth]{analysis/cardio/cufflinks-pca.pdf}
\includegraphics[width=0.5\textwidth]{analysis/cardio/pm-rsem-pca.pdf}
\includegraphics[width=0.5\textwidth]{analysis/cardio/ml-rsem-pca.pdf}
\caption{The first two principal components from principal component analysis
applied to transcript expression estimates using a variety of methods.}
\label{fig:cardiopca}
\end{figure}

\section{Notes on measuring agreement between transcript expression estimates}

Benchmarks in this paper rely heavily on quantifying the agreement between
expression estimates and qPCR, spike-in controls, estimates from other samples,
or ground truth (in the case of simulations). As discussed in Section
\ref{results}, our methodology uses proportionality correlation
\cite{Lovell:2015il} as an alternative to more common measures of correlation or
aggregate error.

In practice, the distribution of transcript expression values tends to be skewed
right with a very low median. This can lead to misleading comparisons in two
ways. First, similarity/dissimilarity measurements like Pearson's correlation or
root mean square error may be dominated by a small number of very highly
expressed transcripts. Second, and conversely, measurements like Spearman's rank
correlation may be dominated by unexpressed, or very low-expression transcripts
which often make up the large majority of transcripts in a relatively
well-annotated species like human.

This problem is not adressed by using median relative difference (MRD), as was used to
argue the accuracy of Kallisto \cite{Bray:2015uj}. Though it may seem robust due
to the use of a median, it does not account for the compositional nature of
RNA-Seq abundance estimates. A large error in one highly expressed transcript
can inflate the relative error in all other transcripts, thus driving up the
median. Furthermore, relative error can be unintuitive in the presense of
zero estimates. To see this, consider that the relative difference
between 0 and $\epsilon$ for an arbitrary positive $\epsilon$ is $\frac{\epsilon
- 0}{(\epsilon + 0)/2} = 2$. Thus, MRD unintuitively treats ``absent versus
barely present'' to be equivalent to ``absent versus highly expressed''.

For these reasons, we believe proportionality correlation is a more appropriate
measure when evaluating relative expression estimates.  The prevailing practical
issue with proportionality correlation is that it is defined only over positive
numbers. Methods that use maximum likelihood (or, depending on the choice of
priors, posterior maximum) estimates do report zeros.  We account for this by
adding a small constant ($c = 0.1 \text{TPM}$) to all expression values before
computing proportionality correlation.


\subsection{Effects of varying additive constant on computed proportionality
            correlation}\label{additive-constant}

As our analysis relies on adding a seemingly arbitrary constant $c$ to
expression values before computing correlation, a natural question is how
sensitive our results our to its precise value. In short, we believe it is quite
insensitive, with results largely consistent when $c$ is varied across several
orders of magnitude (Supplementary Figure \ref{fig:propcor-constant}).

\begin{figure*}
    \centering
    \begin{subfigure}[b]{0.8\textwidth}
    \centering
    \includegraphics[width=0.8\textwidth]{pc-constant-seqc-primepcr.pdf}
    \caption{}
    \end{subfigure}

    \begin{subfigure}[b]{0.8\textwidth}
    \centering
    \includegraphics[width=0.8\textwidth]{pc-constant-seqc-transcript-expression-consistency.pdf}
    \caption{}
    \end{subfigure}

    \caption{To account for zeros and near-zero values, a small constant $c$ is
        added to expression values before computing proportionality correlation.
        Throughout the paper we use $c = 0.1 \text{TPM}$. Here we varied the
        value of $c$ across several orders of magnitude and recomputed
        proportionality correlation in two of the benchmarks used in the paper:
        (a) comparison to PrimePCR qPCR and (b) consistency of transcript
        expression in SEQC data. Though the computed correlation changes
        depending on the value of $c$, methods largely maintain the same order.
    }
    \label{fig:propcor-constant}
\end{figure*}


Given the insensitivity, we claim that $c = 0.1$ reasonable choice. If we assume
between $10^5$ to $10^6$ mRNA are present in a mammalian cell
\cite{Islam:2014is}, then adding $0.1 \text{TPM}$ can be interpreted as
negligibly inflating expression estimates by 0.01 to 0.1 copies per cell.

\subsection{Select analyses repeated with Spearman's rank correlation}\label{spearman-reanalysis}

While we believe proportionality correlation is a more meaningful measure, our
results remain qualitatively similar using Spearman's rank correlation. In this
section we reproduce tables from the main paper using this somewhat more
conventional approach.


\begin{figure*}
\begin{subtable}{0.5\linewidth}
\caption{Comparison of estimates to TaqMan qPCR}
\begin{tabular}[c]{@{}lrrrr@{}}
\toprule\addlinespace
Method & A & B & C & D
\\\addlinespace
\midrule
Isolator & \textbf{0.903} & \textbf{0.892} & \textbf{0.878} & \textbf{0.859}
\\\addlinespace
eXpress & 0.897 & 0.885 & 0.869 & 0.849
\\\addlinespace
Cufflinks & 0.897 & 0.886 & 0.869 & 0.849
\\\addlinespace
Salmon  &  0.895 & 0.881 & 0.868 & 0.845
\\\addlinespace
RSEM/ML  & 0.895 & 0.880 & 0.867 & 0.844
\\\addlinespace
Kallisto &  0.890 & 0.873 & 0.862 & 0.837
\\\addlinespace
BitSeq  &  0.886 & 0.874 & 0.862 & 0.842
\\\addlinespace
Sailfish &  0.875 & 0.843 & 0.842 & 0.804
\\\addlinespace
RSEM/PM  & 0.886 & 0.873 & 0.862 & 0.840
\\\addlinespace
\bottomrule
\addlinespace
\end{tabular}
\label{table:taqman:spearman}
\end{subtable}
\begin{subtable}{0.5\linewidth}
\caption{Comparison of estimates to PrimePCR qPCR}
\begin{tabular}[c]{@{}lrrrr@{}}
\toprule\addlinespace
Method & A & B & C & D
\\\addlinespace
\midrule
Isolator  & \textbf{0.872} & \textbf{0.863} & \textbf{0.837} & \textbf{0.844}
\\\addlinespace
Cufflinks & 0.864 & 0.855 & 0.826 & 0.835
\\\addlinespace
eXpress   & 0.864 & 0.853 & 0.825 & 0.833
\\\addlinespace
Salmon    & 0.858 & 0.846 & 0.821 & 0.825
\\\addlinespace
RSEM/ML   & 0.856 & 0.845 & 0.819 & 0.823
\\\addlinespace
BitSeq    & 0.855 & 0.842 & 0.821 & 0.827
\\\addlinespace
Kallisto  & 0.855 & 0.840 & 0.817 & 0.821
\\\addlinespace
Sailfish  & 0.840 & 0.811 & 0.797 & 0.793
\\\addlinespace
RSEM/PM   & 0.850 & 0.838 & 0.815 & 0.819
\\\addlinespace
\bottomrule
\addlinespace
\end{tabular}
\label{table:primepcr:spearman}
\end{subtable}

\begin{subtable}{0.5\linewidth}
\caption{Comparison of estimates to ERCC spike-in controls}
\begin{tabular}[c]{@{}lrrrr@{}}
\toprule\addlinespace
Method & A & B & C & D
\\\addlinespace
\midrule
Isolator  & \textbf{0.974} & \textbf{0.973} & \textbf{0.975} & \textbf{0.976}
\\\addlinespace
Salmon    & 0.969 & 0.968 & 0.966 & 0.966
\\\addlinespace
Kallisto  & 0.964 & 0.964 & 0.961 & 0.965
\\\addlinespace
Sailfish  & 0.968 & 0.961 & 0.959 & 0.959
\\\addlinespace
Cufflinks & 0.964 & 0.964 & 0.961 & 0.964
\\\addlinespace
RSEM/PM   & 0.950 & 0.946 & 0.951 & 0.948
\\\addlinespace
RSEM/ML   & 0.945 & 0.950 & 0.948 & 0.948
\\\addlinespace
BitSeq    & 0.941 & 0.938 & 0.945 & 0.942
\\\addlinespace
eXpress   & 0.937 & 0.944 & 0.941 & 0.942
\\\addlinespace
\bottomrule
\addlinespace
\end{tabular}
\label{table:ercc:spearman}
\end{subtable}
\begin{subtable}{0.5\linewidth}
\caption{Consistency of SEQC sample estimates}
\begin{tabular}[c]{@{}lrr@{}}
\toprule\addlinespace
Method & $c$ vs $0.75a + 0.25b$ & $d$ vs $0.25a + 0.75b$
\\\addlinespace
\midrule
Isolator  & \textbf{0.972} & \textbf{0.972}
\\\addlinespace
BitSeq    & 0.946 & 0.947
\\\addlinespace
RSEM/PM   & 0.936 & 0.935
\\\addlinespace
Sailfish  & 0.882 & 0.877
\\\addlinespace
RSEM/ML   & 0.841 & 0.838
\\\addlinespace
Salmon    & 0.823 & 0.821
\\\addlinespace
Kallisto  & 0.845 & 0.842
\\\addlinespace
eXpress   & 0.868 & 0.867
\\\addlinespace
Cufflinks & 0.865 & 0.864
\\\addlinespace
\bottomrule
\addlinespace
\end{tabular}
\label{table:seqctranscripts:spearman}
\end{subtable}

\caption{
Results from Figure 1 from the main paper recomputed using Spearman's rank correlation.
}
\end{figure*}


\begin{table}
\begin{center}
\label{table:rlsim:spearman}
\begin{tabular}[c]{@{}ll@{}}
\toprule\addlinespace
Method & Correlation
\\\addlinespace
\midrule
Isolator   & 0.724
\\\addlinespace
Kallisto   & 0.727
\\\addlinespace
Salmon     & \textbf{0.728}
\\\addlinespace
RSEM/ML    & 0.712
\\\addlinespace
Cufflinks  & 0.636
\\\addlinespace
eXpress    & 0.605
\\\addlinespace
Sailfish   & 0.648
\\\addlinespace
RSEM/PM    & 0.605
\\\addlinespace
BitSeq     & 0.595
\\\addlinespace
\bottomrule
\addlinespace
\end{tabular}
\caption{Table 1 from the main paper recomputed using Spearman's rank correlation.}
\end{center}
\end{table}

\bibliographystyle{plainnat}
\bibliography{references,extra-references}

\end{document}
